### PR TITLE
Support passing logits processor to MLXModel

### PR DESCRIPTION
**Pull request by g-eoj**
Enables the model to produce structured output through grammar or regex.

Adds a base class that is meant to provide a structure for using logits processor libraries with MLXModels.

  Example:
  ```python
  import outlines
  from smolagents import CodeAgent, BaseMLXLogitsProcessor, MLXModel


  class RegexLogitsProcessor(BaseMLXLogitsProcessor):
      def __init__(self, grammar, tokenizer):
          self._outlines_processor = outlines.processors.RegexLogitsProcessor(
              regex_string=grammar,
              tokenizer=outlines.models.TransformerTokenizer(tokenizer)
          )
      def __call__(self, input_ids, logits):
          processed_logits = self._outlines_processor(
              input_ids,
              logits.reshape(-1)
          )
          return processed_logits.reshape(1, -1)

  model = MLXModel(
      model_id="mlx-community/Qwen2.5-Coder-32B-Instruct-4bit",
      max_tokens=5000,
      logits_processor=RegexLogitsProcessor
  )

  CodeAgent(
      model=model,
      grammar=r'Thought: ([^\\.\n]+?\\.){1,3}\nCode:\n```(?:py|python)?\n[^`]+?```<end_code>',
      tools=[],
  )
  ```

### CODE REVIEW

**Comment by g-eoj** [comment]
Question for anyone interested: should the logits processor just get passed directly to the `__call__` method? So it'd end up like this:

```
model = MLXModel(
    model_id="mlx-community/Qwen2.5-Coder-32B-Instruct-4bit",
    max_tokens=5000
)
CodeAgent(
    model=model,
    logits_processor=RegexLogitsProcessor(
        r'Thought: ([^\\.\n]+?\\.){1,3}\nCode:\n```(?:py|python)?\n[^`]+?```<end_code>'
    ),
    tools=[],
)
```

In retrospect I think this approach is more flexible. Originally I didn't want to recompile the processor grammar/regex every call but I discovered that has to happen anyways to reset the logits processor.
